### PR TITLE
Prefer update URLs over PluginURI in My Subscriptions

### PR DESCRIPTION
**Pull request by vishnugopal**
### Submission Review Guidelines:

- [x]  I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [x]  I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- [x] I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).

### Changes proposed in this Pull Request:

Plugins in My Subscriptions that don't have a linked subscription has a product URL that is currently populated using the plugin's `PluginURI` metadata. This however often results in outdated product URLs (especially if you haven't updated the plugin for a long time).

In this screenshot for example, WooCommerce Customer/Order/Coupon Export has an extra and unnecessary `www` in the URL that is fetched from the plugin's `PluginURI`.

CleanShot 2024-05-29 at 16 28 22

This PR changes the `product_url` to the URL returned from woocommerce.com's `update-check` endpoint. If such a URL is unavailable, it always falls back to the existing `PluginURI` behaviour.

Related to 20529-gh-Automattic/woocommerce.com

### How to test the changes in this Pull Request:

Let's first reproduce the bug:

1. Use a site [like JN](https://jurassic.ninja/) to create a Woo site with latest version.
2. Install and activate a plugin for which you don't have a purchase on woocommerce.com. Note that this plugin also needs an outdated `PluginURI` field in the ZIP. For A8C reviewers, you can download `Customer/Order/Coupon CSV Import Suite` from [here](https://woocommerce.com/my-account/downloads/).
3. Make sure you don't have a purchase for  `Customer/Order/Coupon CSV Import Suite` .
4. Connect the site to woocommerce.com from WooCommerce > Extensions > My Subscriptions (`/wp-admin/admin.php?page=wc-admin&tab=my-subscriptions&path=%2Fextensions`).
5. Now, you should see the wrong plugin URL (with `www`) when hovering over the product listing.

Now, test the fix:
1. Use a site like JN to create a site.
2. Install [woocommerce.zip](https://github.com/woocommerce/woocommerce/files/15484015/woocommerce.zip) from this branch.
3. Follow steps 2-4 above.
4. You should now see the correct URL:

CleanShot 2024-05-29 at 17 15 06

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

Prefer update URLs over PluginURI in My Subscriptions for plugins without a subscription.

#### Comment 

</details>

### CODE REVIEW

**Comment by vishnugopal** [comment]
Closing and reopening to re-trigger CI.

File: plugins/woocommerce/includes/admin/helper/class-wc-helper.php [hunk 1]
```diff
@@ -1598,11 +1598,12 @@ public static function get_subscription_list_data() {
 			);
 		}
 
+		// Fetch updates so we can refine subscriptions with information about updates.
+		$updates = WC_Helper_Updater::get_update_data();
```
Comment (denho) [inline plugins/woocommerce/includes/admin/helper/class-wc-helper.php:1611]: Props for moving this call out of the loop.

File: plugins/woocommerce/includes/admin/helper/class-wc-helper.php [hunk 2]
```diff
@@ -1613,6 +1614,11 @@ public static function get_subscription_list_data() {
 			if ( ! empty( $updates[ $subscription['product_id'] ] ) ) {
 				$subscription['version'] = $updates[ $subscription['product_id'] ]['version'];
 			}
+
+			// If the update endpoint returns a URL, we prefer it over the default PluginURI.
+			if ( ! empty( $updates[ $subscription['product_id'] ]['url'] ) ) {
+				$subscription['product_url'] = $updates[ $subscription['product_id'] ]['url'];
+			}
```
Comment (denho) [inline plugins/woocommerce/includes/admin/helper/class-wc-helper.php:1631]: This works, but overwriting the previously declared value complicates this code a bit. It becomes harder to track how the value was set.

I'm curious if you considered removing this second `foreach` loop, as it seems redundant given it iterates the same subscriptions that are defined in the loop above?

**Review by denho — APPROVED** [review]
Thank you, @vishnugopal. Tested and works for me. Left non-blocking comments.

File: plugins/woocommerce/includes/admin/helper/class-wc-helper.php (hunk 2 above)
Comment (vishnugopal) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper.php:1631]: Hi @denho, the code is a bit unobvious, but the function combines data from 2 sources: connected subscriptions and locally installed Woo products. The connected subscriptions [are fetched here](https://github.com/woocommerce/woocommerce/pull/47950/files#diff-32b26bf8430b272079f3c981317024f2a7d7848b4198ae7878716cc133d10961R1567), and then locally installed Woo products [are added here](https://github.com/woocommerce/woocommerce/pull/47950/files#diff-32b26bf8430b272079f3c981317024f2a7d7848b4198ae7878716cc133d10961R1587). 

The [second loop](https://github.com/woocommerce/woocommerce/pull/47950/files#diff-32b26bf8430b272079f3c981317024f2a7d7848b4198ae7878716cc133d10961R1614) iterates over this merged subscriptions list, and further refines it. I couldn't find a way to make this clearer in code, so I added some (hopefully) descriptive comments.

**Comment by vishnugopal** [comment]
Adding in Proton as a reviewer to help merge this, thank you! 🙇🏽

**Review by naman03malhotra — COMMENTED** [review]
👍 ✅

**Pull request merged**